Add optional param to the identity client and api

kissmikijr opened the pull request:
## Hey, I just made a Pull Request!

continuation of: https://github.com/backstage/backstage/pull/14533

add optional parameter to the identity client and API

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

punkle commented on plugins/auth-node/src/DefaultIdentityClient.test.ts:
```diff
@@ -401,17 +401,29 @@ describe('DefaultIdentityClient', () => {
+        client.getIdentity({
+          request: { headers: {} },
+        } as IdentityApiGetIdentityRequest),
+      ).rejects.toThrow('Identity was not provided');
```
Should we assert that this has thrown an AuthenticationError so that the router knows what to return to the http response.

punkle commented on plugins/auth-node/src/DefaultIdentityClient.test.ts:
```diff
@@ -401,17 +401,29 @@ describe('DefaultIdentityClient', () => {
+          optional: true,
         } as IdentityApiGetIdentityRequest),
-      ).toEqual(undefined);
+      ).resolves.toEqual(undefined);
```
👍

punkle commented on plugins/auth-node/src/DefaultIdentityClient.test.ts:
```diff
@@ -401,17 +401,29 @@ describe('DefaultIdentityClient', () => {
       await expect(
         client.getIdentity({
-          request: { headers: { authorization: `Bearer bad-token` } },
+          request: { headers: { authorization: 'Bearer bad-token' } },
```
Im aware that I probably wrote this expectation. But I wonder is it possible to check the error type to make sure its an AuthenticationError?

punkle commented on plugins/playlist-backend/src/service/router.ts:
```diff
@@ -85,7 +85,7 @@ export async function createRouter(
     );
 
-    const user = await identity.getIdentity({ request });
+    const user = await identity.getIdentity({ request, optional: true });
```
This comment is pedantic and can be happily ignored.

Im wondering if we should decide on a name to call the response from this call and use it across the other plugins too.

We are already using: user, userIdentity, callerIdentity and identity.

kissmikijr commented on plugins/auth-node/api-report.md:
```diff
@@ -43,6 +44,10 @@ export function getBearerTokenFromAuthorizationHeader(
 
 // @public
 export interface IdentityApi {
+  getIdentity(
+    options: Omit<IdentityApiGetIdentityRequest, 'optional'>,
+  ): Promise<BackstageIdentityResponse>;
+  // (undocumented)
```
@Rugvip I tried the overload that you mentioned here https://github.com/backstage/backstage/pull/14533#discussion_r1040172973
Not sure I was successful, haha so please check if this is what you wanted to achieve with it.

Rugvip replied:
Not quite as I don't think this'd catch `optional: false`. Cleanest way is probably a type parameter for the request.

```ts
interface IdentityApiGetIdentityRequest<TOptional extends boolean = boolean> {
  optional: TOptional,
  ...
}

getIdentity(options: IdentityApiGetIdentityRequest<true>): Promise<BackstageIdentityResponse | undefined>;
getIdentity(options: IdentityApiGetIdentityRequest): Promise<BackstageIdentityResponse>;
```

kissmikijr commented on plugins/permission-backend/src/service/router.test.ts:
```diff
@@ -62,6 +66,28 @@ describe('createRouter', () => {
+          token,
+        });
+      },
+    ) as any;
```
I added this any cast here because tsc wasn't passing because it assumed this function will be the other signature of the getIdentity method. Not sure how to do it properly , I wasn't really able to figure out how to mock an overloaded function in jest.

kissmikijr commented on plugins/scaffolder-backend/src/service/router.ts:
```diff
@@ -86,62 +92,77 @@ function isSupportedTemplate(entity: TemplateEntityV1beta3) {
 
-      if (!header) {
-        return undefined;
+class CustomIdentityClient implements IdentityApi {
```
I needed to convert this custom implementation of the IdentityApi to a class to be able to properly implement the overloaded function for it.

kissmikijr commented:
yaay the build passed finally! :) @Rugvip I think this is ready for another round of reviews, this addresses most of the comments from this review https://github.com/backstage/backstage/pull/14533

kissmikijr commented:
Not stale! Waiting for a review!

Rugvip reviewed (changes requested):
Quick comment but also wanted to check in on the urgency of this. Is this blocking something on your side, or just a bit more maintenance? Reason I'm asking is that I would prefer to take a more broader look at auth through #15999, but we're not there quite yet still
